service/storage_service.cc

```diff
@@ -2070,20 +2070,34 @@ future<> storage_service::decommission() {
                     slogger.info("decommission[{}]: became a group 0 non-voter", uuid);
                 }
 
-                // Step 6: Leave the token ring
+                // Step 6: Verify that other nodes didn't abort in the meantime.
+                // See https://github.com/scylladb/scylladb/issues/12989.
+                req.cmd = node_ops_cmd::query_pending_ops;
+                parallel_for_each(nodes, [&ss, req, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                        slogger.debug("decommission[{}]: Got query_pending_ops response from node={}, resp.pending_ops={}", uuid, node, resp.pending_ops);
+                        if (boost::find(resp.pending_ops, uuid) == resp.pending_ops.end()) {
+                            throw std::runtime_error(format("decommission[{}]: Node {} no longer tracks the operation", uuid, node));
+                        }
+                        return make_ready_future<>();
+                    });
+                }).get();
+
+                // Step 7: Leave the token ring
                 slogger.info("decommission[{}]: leaving token ring", uuid);
                 ss.leave_ring().get();
                 left_token_ring = true;
                 slogger.info("decommission[{}]: left token ring", uuid);
 
-                // Step 7: Finish token movement
+                // Step 8: Finish token movement
```

Comment:
The change to wait for node has left the cluster in node_ops_cmd::decommission_done handler is not in any release yet.  We can introduce a new node_ops_cmd::query_node_status to wait for a node has left the cluster before decommission() returns instead of in node_ops_cmd::decommission_done.  This way, we can send the decommission_done before advertise the tokens, as we do for other node ops, including removenode ops. 

In removenode(), we can use the new node_ops_cmd::query_node_status to make sure all nodes see the removed node is not part of the ring before we return removenode too.

Reply:
Sounds promising. But what about this:

```
            _db.invoke_on_all([](replica::database &db) {
                for (auto& table : db.get_non_system_column_families()) {
                    table->trigger_offstrategy_compaction();
                }
            }).get();
```

Is it really needed as part of decommission_done? Other node ops don't have it.

Reply:
We can keep it as it is. Triggering off-strategy is needed because the leaving node pushes data to the nodes that will stay in the cluster. They have no idea when the push will finish.

Reply:
Let's do the decommission_done split incrementally, since it's not strictly necessary for improving safety.

Reply:
@asias, any objections for merging as-is?

Reply:
OK. No problem. We can have a follower up after this.